src/htm/encoders/SimHashDocumentEncoder.cpp

```diff
@@ -0,0 +1,276 @@
+ * Copyright (C) 2016, Numenta, Inc. https://numenta.com
+ *               2019, David McDougall
+ *               2019, Brev Patterson, Lux Rota LLC, https://luxrota.com
+ *
+ * This program is free software: you can redistribute it and/or modify it
+ * under the terms of the GNU Affero Public License version 3 as published by
+ * the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE. See the GNU Affero Public License for
+ * more details.
+ *
+ * You should have received a copy of the GNU Affero Public License along with
+ * this program.  If not, see http://www.gnu.org/licenses.
+ * -------------------------------------------------------------------------- */
+
+/** @file
+ * SimHashDocumentEncoder.cpp
+ * @since 0.2.3
+ */
+
+#include <bitset>
+#include <climits> // CHAR_BIT
+
+#include <hasher.hpp> // digestpp: sha3+shake256 hash digests
+#include <algorithm/sha3.hpp>
+#include <algorithm/shake.hpp>
+
+#include <htm/encoders/SimHashDocumentEncoder.hpp>
+
+
+namespace htm {
+
+  /**
+   * Constructor
+   *
+   * @see SimHashDocumentEncoder.hpp
+   */
+  SimHashDocumentEncoder::SimHashDocumentEncoder(const SimHashDocumentEncoderParameters &parameters)
+  {
+    initialize(parameters);
+  }
+
+  /**
+   * Initialize
+   *
+   * @see SimHashDocumentEncoder.hpp
+   */
+  void SimHashDocumentEncoder::initialize(const SimHashDocumentEncoderParameters &parameters)
+  {
+    // Pre-processing, Check parameters
+    NTA_CHECK((parameters.activeBits > 0u) || (parameters.sparsity > 0.0f))
+      << "Need only one argument of: 'activeBits' or 'sparsity'.";
+    NTA_CHECK(parameters.size > 0u)
+      << "Missing 'size' argument.";
+
+    // Make local copy of arguments for processing
+    args_ = parameters;
+
+    // process: handle 'sparsity' param instead of 'activeBits' param
+    if(args_.sparsity > 0.0f) {
+      NTA_CHECK((args_.sparsity >= 0.0f) && (args_.sparsity <= 1.0f))
+        << "Argument 'sparsity' must be a float in the range 0.0-1.0.";
+      NTA_CHECK(args_.size > 0u)
+        << "Argument 'sparsity' requires that the 'size' also be given.";
+      args_.activeBits = (UInt) round(args_.size * args_.sparsity);
+    }
+
+    // Post-processing, Sanity check the parameters
+    NTA_CHECK(args_.size > 0u);
+    NTA_CHECK(args_.activeBits > 0u);
+    NTA_CHECK(args_.activeBits < args_.size);
+
+    // Initialize parent class with finalized params
+    BaseEncoder<std::vector<std::string>>::initialize({ args_.size });
+  } // end method initialize
+
+  /**
+   * Encode (Main calling style)
+   *
+   * If param "tokenSimilarity" is passed, we'll additionally loop through
+   *  all the letters in all the tokens.
+   * Each token/letter will be hashed with SHA3+SHAKE256 to get
+   *  a varible-length (param "size") binary digest output. These vectors will
```

Comment:
what is with this "variable-len size"? I thought the encoder does have a fixed encoding size